### ISSUE TITLE
Install Canvas, learn how plugins work, and install/adapt analytics collection system

### ISSUE THREAD

**Issue by science** [learningtapestry/analytics-ruby-client#3]
- [ ] Install Canvas on your local machine and generally get familiar with it
- [ ] Understand how plugins work - add info to #2 as you get more knowledge and develop strategies
- [ ] Talk with @science and @jasonhoekstra about analytics tool to understand how it works / what it is for
- [ ] Develop a strategy for using the analytics system with Canvas / review with @science

To be continued in this or other tickets once we get that far..

**Comment by aspino** [comment]
@science: I've already set up the Canvas LMS app on my local environment and it's running nicely. I followed the manual install so I could knew exactly the requirements and steps involved. However, they also provide a one-time automated script that performs the installation on all major systems. 

Considering that:
* Maybe there's no need to waste time building a docker/vagrant image, because the development environment is relatively easy to set-up when using the automated script. Besides, there is already a project that provides a docker image, in case somebody wants it: https://github.com/playlyfe/docker-canvas. It seems a little outdated, but updating it to the latest version should be easy enough.
* The production version could really benefit from something like docker, so the db/web server/redis setup is totally automated and ready to deploy on the servers whenever we want. However, I'm not familiar enough with ansible, so maybe this is something already covered by the service. By the way, I'm assuming the future demo version will run using the production settings. If that's not the case, you can forget everything I said :-)

Ok, I think I'm going to continue learning about the plugins. It would be interesting to know what were the difficulties that the previous developer encountered while using/studying the canvas plugins. Does anyone know? or maybe he left some documentation besides #2 ?

**Comment by science** [comment]
Thanks. For now we aren't planning a production installation of Canvas: we want to build a demo LMS that can showcase our product and also have a good understanding of Canvas so we can recommend it to clients, and/or implement it for them along with our expansion of the Canvas system.

Probably our Ansible installer could handle installing Canvas directly if there is a set of install scripts. Or we could deploy the docker container via Ansible and run it that way. But don't worry about production deployment - a manual install onto a demo server will be all that we need coming up (with set up instructions).

I'm not sure exactly what difficulties the developer had - digging into the code base might have been too intricate. Or possibly figuring out how to make our unusual plugin work with Canvas.

Keep in mind there are two things we want to accomplish with the plugin:

* Use the Analytics Ruby Client library to display dashboard content: we think this will be easy
* Use the Analytics Javascript (or possibly Ruby lib similar to NewRelic) library to perform collection of activity data from Canvas. Big question: Does the plugin architecture of Canvas allow any use along these lines or are we monkey patching or something else?

**Comment by aspino** [comment]
My first impression on the plugins system is that they're not plugins in a traditional meaning. They have some common mechanism to register them, as well as define the partials and validations used in the plugin configuration UI, but other than that I don't think they follow a unified interface to interact with the core system. I've not been able (yet) to find some kind of common interface or adapter classes that could imply that the core system is ready to be called via an external component.

The plugins in the application are mostly distributed as internal gems, but then there has to be some parts in the core application (controllers, models, etc.) that call these plugin classes explicitly. To me this is not a typical plugin architecture, just some way to organize the code in gems and define a unified interface for its configuration. Besides, it doesn't follow the Open-Closed Principle, because, as far as I can tell, the system has to know about the plugins in order to use them, so the only way to actually write a plugin is modifying the code in the core system to make it aware of the new options/features.

Let me give you a couple of examples:

* The *diigo* plugin (inside _gems/diigo_) defines the _get_bookmarks_ method (https://github.com/instructure/canvas-lms/blob/stable/gems/diigo/lib/diigo/connection.rb#L60) as part of the plugin. Then, the _UsersController_ class (which I understand is part of the core) calls this method explicitly when it's creating a user service (https://github.com/instructure/canvas-lms/blob/stable/app/controllers/users_controller.rb#L937).
* Many other plugins, such as Google Docs, LinkedIn or Twitter Integrations, are also distributed in its own gems, but then in the _ApplicationController_ class (https://github.com/instructure/canvas-lms/blob/stable/app/controllers/application_controller.rb#L1424) there's a switch statement that controls what plugins are enabled. There are also many other parts in the code (custom initializers, models, controllres, etc.) that reference the plugin names or its methods explicitly.

It's possible that other parts of the system are actually designed to be used under a common interface or via pluggable components.  Plugins like the _cisco_webex_ seem to be a little more detached (they are distributed as an external gem), but I couldn't find any code that would make me think that they use any particular plugin interface.

So, I could be wrong, but I'm pretty sure all sensible options to write a canvas LMS plugin involve modifying the app to include the desired features, then open a pull request and ask for the inclusion in the project code.

**Comment by science** [comment]
Thank you. This actually makes some business sense (for Instructure), and would explain why Alter (the first programmer who looked at this) had trouble. He probably couldn't believe what he was seeing. :)

It sounds like this system allows the company to monitor and control who is integrating with a non-forked version of Canvas. They claim to be very open on this front, so maybe the next step is to design our integration strategy and post an issue on their ticket system to propose the integration and see what they say.

Possibly we should talk sync to develop a design? Or do you feel like you understand what we need? Basically I think we need to accomplish two and half things:
1. We want to inject some javascript into outbound pages to allow us to place monitoring JS code on all LMS pages.
1.1. We need to track user identity during injection so we can include user identity info to our monitoring code.
1 1/2. We want to hook/monitor some controller, model and similar events to track some activities on the LMS side of the system. Similar to \#1 we want to send these events back to our Learning Record Store (LRS).
2. We want to provide some easy tools for building dashboard displays of our data, using this analytics-ruby-client or similar, or possibly some drop-in controller/pages for Rails that do most of this automatically.

**Comment by aspino** [comment]
I've been thinking about the cleanest way to perform the integration now that we have a better understanding of the way plugins work and, since most of the tasks that we need to accomplish seem related to tracking data and sending it to Learning Tapestry servers, I think we could consider implementing it via a Rack middleware (as an external gem preferably), similar to the way gems like newrelic_rpm or rack-livereload work. This approach would give us these benefits:

* A total separation from the Canvas LMS application. If we're lucky, we'd only have to ask them to add the gem entry in the app's Gemfile and, perhaps, add some configuration file or the necessary declarations in their plugin system. The integration could be switchable, for those customers that don't want to enable it.
* We'd be much less exposed to future changes in the Canvas LMS application itself that could break our code, and viceversa. Besides, our development cycle would not be tied to theirs, we could decide when to release new versions of the gem on our own.
* The tracking javascript code could be injected dynamically in the pages we're interested in. Since you said they might consider this to be a very invasive change, this way the integration would be transparent and, I hope, more welcomed.
* I think the dashboard displays of our data could also be injected, by using js or other mechanism, if we feel we like this approach and it makes sense.

Of course, this kind of integration means we're working at the rack level, not inside the application, and would be more difficult for us to implement, but I think the potential benefits and separation from the Canvas app can be worth it.

@science Let me know what you think. If you want, I could start doing some proof of concept locally and show you a simple dummy integration once I'm done.

**Comment by science** [comment]
I think I understand the proposal - but let me ask some clarifying questions. If implemented in Rack, could this solution be equally easily be built using a generic proxy (or reverse proxy) server? Or for example, could it be written into Nginx directly and rewrite/inject the javascript from there?

I think this would be an html streaming solution, which creates some challenges. But it might be relatively easy to wait for the </body> tag and inject the javascript tag right above it?

Am I following along with the design strategy?

@jasonhoekstra has played with writing Lua plugins to Nginx, and this type of plugin is very, very fast, compared to Ruby/Rack. It might be a good spot for this. And if it turns out some people run Canvas on Apache, we'd just have to port the tech there.

On top of that, this solution would work for *any* provider running nginx. They'd just have to plugin our Lua module into their Nginx server and get analytics on all their pages.

Are we talking about the same ideas?

**Comment by aspino** [comment]
Yes, I think it's basically the same idea: inject some custom js/html into the response. The difference is the level at which this injection is performed, either at web server/proxy level (nginx or apache) or at application level (rack middleware). And sure, just injecting the javascript should not be difficult, no matter the way you do it.

I guess what you propose could be done with something like ngx_http_sub_module (http://nginx.org/en/docs/http/ngx_http_sub_module.html), or, of course, writing a Lua plugin. For Apache I guess some similar module will exist, or you can also resort to Lua scripting.

I think both approaches have their merits. The nginx/apache solution is 100% independent of the served application (so no need to coordinate with the Canvas LMS developers) and, as you say, it will probably be way faster than using a rack middleware. Besides, it would be useful for any kind of application, not just the Canvas LMS, so maybe it's interesting for your company to develop such a tool.

However, the rack middleware approach is closer to the rails stack, and that means it could use any ruby class belonging to the application in case it needs to. I'm also thinking about cases where the web server does not exist, like heroku, that serve the application straight from the app server, or when the client uses a web server different from nginx or apache. The rack solution wouldn't need to consider different combinations, it would work fine for every rack application (sinatra, rails, padrino, etc.).

Personally, if I'd have to implement it only for the Canvas app, I think I'd choose the rack solution, but maybe just because I'm a ruby application developer and I'm more inclined to implement the solutions at the application level rather than at the infrastructure level. But if your idea is closer to allow hooking your analytics module into any kind of application/technology, maybe an nginx plugin would be more appropriate. 

Well, I hope you find my feedback interesting. I must say that my nginx skills are average and I have no experience developing LUA plugins, but I'm always happy to learn new things :-), so let me know what you think.

**Comment by science** [comment]
Yeah - there's one issue I haven't mentioned that probably directs us towards the Rack approach. The javascript library needs the user ID on the page or in the javascript require statement. Our analytics system doesn't work without some unique ID for the user who is on the page.

If we are in Rack, we might be able to get that data from Canvas? Otherwise, we can do the injection but the application developer would still have to provide us the user id metadata on each page.

**Comment by aspino** [comment]
Well, now that you mention it, I've seen that Canvas sets some headers in the response, and one of them seems to be the user-id so, theoretically, we could just read that http header, either in rack or nginx. Take a look at the screenshot:

![screen shot 2015-07-29 at 19 01 42](https://cloud.githubusercontent.com/assets/1306310/8964194/79a8705e-3624-11e5-9184-5b3ae95fd828.png)

However, if the user-id we need is the actually the e-mail or any other field, then I'm afraid we'll need direct access to the application in order to retrieve those fields. 

I have a question: how does the analytics system know who is really the user if it only receives a unique id from Canvas? I guess it will perform some kind of mapping between the user-id configured in the analytics API and the user-id from the Canvas app, so that it knows that all received data from some given canvas-user-id belongs to some user in the analytics system. Is that right or I'm talking nonsense :-D?

**Comment by science** [comment]
Well that's interesting. Seems awfully thoughtful of them, if it's used reliably.

You're not talking nonsense - our design is not obvious, and you may see flaws in it - so input welcome.

Our analytics system works on an opaque user identifier model. We never authenticate a user, we just "trust" that the data coming from the system is reliable. So it's easy to spoof data, as it is with all javascript reporting solutions. E.g., AFAIK it's easy to send junk data to a Google Analytics customer ID.

The crucial design principle in analytics is that whatever you send us for user id during collection is what you have to send us as user id when using our reporting api. As long as it's the same, we're a "dump pipe" regarding what that identity means.

Does that seem right/ok to you?

**Comment by aspino** [comment]
Yes, I understand, and following the Google Analytics approach seems like a very sensible choice to me.

But there's something that I think I still don't understand... how do you prevent having duplicate user IDs in your analytics system? I mean, in the case of GA they generate a unique user id for their system, and then you use it to configure the js snippet and send data. But in your case I understand that you ask for the user to provide its own user ID, for instance taking the user ID from the canvas app or otherwise. How do you prevent then that two users end up choosing the same ID?

**Comment by science** [comment]
Each submitting org has an api key. Does that answer your question?

**Comment by aspino** [comment]
Ah, of course! :-D

**Comment by aspino** [comment]
Quick update: 

I did a quick proof of concept setting up nginx as a reverse proxy + ngx_http_sub_module, and things work as expected. The JS code is injected in the `<head>` tag and executed as usual. Inside the JS code I perform a HEAD request because I'm only interested in the headers, then I log in the console the obtained Canvas User ID.

Here's the code in case you also want to try it for yourself (it binds to port 8081 and assumes your app is running on port 3000):

```
http {
  upstream canvas_app
  {
      server localhost:3000;
  }
  server
  {
      listen                  8081;

      # redirect on errors
      error_page              500 502 504 /50x.html;

      location /
      {
          proxy_set_header        X-Forwarded-For $proxy_add_x_forwarded_for;
          proxy_set_header        Host $http_host;
          proxy_redirect          off;

          if (!-f $request_filename)
          {
              proxy_pass  http://canvas_app;
              break;
          }

          sub_filter      </head>

          '<script type="text/javascript">

          // Obtain API key from some place...
          var apiKey = "YOUR_API_KEY";

          // Perform a HEAD request to retrieve the Canvas User ID contained in the header
          var req = new XMLHttpRequest();
          req.open("HEAD", document.location, false);
          req.send(null);
          var canvasUserId = req.getResponseHeader("x-canvas-user-id");
          console.log("Canvas User ID: " + canvasUserId)

          // We can now submit the data to our Analytics API
          // analyticsApiCall(apiKey, canvasUserId, document.location, ...);
          </script></head>';

          sub_filter_once on;
      }
  }
}
```
Of course, this is just a simple code snippet to prove our point, and it's far from being in a distributable state. We should decide now how to develop further this module, what's the best option to distribute it in a way that's both easy to install and update, what options for retrieving the user id and the API key we should provide to customers, etc.

So, let me know what you think.

**Comment by judell** [comment]
@science and @aspino the discussion of Canvas plugins was very helpful, we (at Hypothes.is) are considering whether/how to head down that path, so thanks! It sounds like maybe you discovered that wasn't where you needed to go, though?
